pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -2559,28 +2684,84 @@ def ResidualWaterRHSRule(model, r, wt, t):
-            model.v_F_TreatedWater[r, t]
-            == sum(model.v_F_Piped[r, l, t] for l in model.s_L if (r, l) in model.s_LLA)
-            + model.v_F_DesalinatedWater[r, t]
+    def TreatedWaterBalanceRule(model, r, t):
+        constraint = model.v_F_TreatedWater[r, t] == sum(
+            model.v_F_Piped[r, l, t]
+            for l in model.s_L
+            if (r, l) in model.s_LLA
+            and model.df_parameters["LLA"][r, l] == TreatmentStreams.treated_stream
+        ) + sum(
+            model.v_F_Trucked[r, l, t]
+            for l in model.s_L
+            if (r, l) in model.s_LLT
+            and model.df_parameters["LLT"][r, l] == TreatmentStreams.treated_stream
         )
         return process_constraint(constraint)
 
-    model.TreatedWater = Constraint(
-        model.s_R, model.s_T, rule=TreatedWaterRule, doc="Treated water balance"
+    model.TreatedWaterBalance = Constraint(
+        model.s_R, model.s_T, rule=TreatedWaterBalanceRule, doc="Treated water balance"
     )
 
-    def BeneficialReuseCapacityRule(model, o, t):
+    # Create a set of all treatment sites for which the residual stream should
+    # be modeled
+    treatment_sites_with_modeled_residual = {
+        org
+        for ((org, dest), value) in list(model.df_parameters["LLA"].items())
+        + list(model.df_parameters["LLT"].items())
+        if org in model.s_R and value == TreatmentStreams.residual_stream
+    }
+
+    def ResidualWaterBalanceRule(model, r, t):
+        # If residual stream from a treatment site should be modeled, then
+        # create constraint for the residual water. Otherwise, skip.
+        if r in treatment_sites_with_modeled_residual:
+            constraint = model.v_F_ResidualWater[r, t] == sum(
+                model.v_F_Piped[r, l, t]
+                for l in model.s_L
+                if (r, l) in model.s_LLA
+                and model.df_parameters["LLA"][r, l] == TreatmentStreams.residual_stream
+            ) + sum(
+                model.v_F_Trucked[r, l, t]
+                for l in model.s_L
+                if (r, l) in model.s_LLT
+                and model.df_parameters["LLT"][r, l] == TreatmentStreams.residual_stream
+            )
+            return process_constraint(constraint)
+        else:
+            return Constraint.Skip
 
+    model.ResidualWaterBalance = Constraint(
+        model.s_R,
+        model.s_T,
+        rule=ResidualWaterBalanceRule,
+        doc="Residual water balance",
+    )
+
+    def BeneficialReuseMinimumRule(model, o, t):
         constraint = (
             sum(model.v_F_Piped[l, o, t] for l in model.s_L if (l, o) in model.s_LLA)
             + sum(
                 model.v_F_Trucked[l, o, t] for l in model.s_L if (l, o) in model.s_LLT
             )
-            <= model.p_sigma_Reuse[o] + model.v_S_ReuseCapacity[o]
+            >= model.p_sigma_ReuseMinimum[o, t]
         )
+        return process_constraint(constraint)
 
+    model.BeneficialReuseMinimum = Constraint(
```

Comment:
def BeneficialReuseFlowRule(model, o, t):
        constraint = (
            sum(model.v_F_Piped[l, o, t] for l in model.s_L if (l, o) in model.s_LLA)
            + sum(
                model.v_F_Trucked[l, o, t] for l in model.s_L if (l, o) in model.s_LLT
            )
        ) <= model.p_M_Flow * model.vb_y_BeneficialReuse[o, t]
        return process_constraint(constraint)

    model.BeneficialReuseZeroFlow = Constraint(
        model.s_O,
        model.s_T,
        rule=BeneficialReuseFlowRule,
        doc="Beneficial reuse zero flow when not selected",
    )

    def BeneficialReuseMinimumFlowRule(model, o, t):
        constraint = (
            sum(model.v_F_Piped[l, o, t] for l in model.s_L if (l, o) in model.s_LLA)
            + sum(
                model.v_F_Trucked[l, o, t] for l in model.s_L if (l, o) in model.s_LLT
            )
        ) >= model.p_sigma_ReuseMinimum[o, t] + 1e-4 - model.p_M_Flow * (
            1 - model.vb_y_BeneficialReuse[o, t]
        )
        return process_constraint(constraint)

    model.BeneficialReuseMinimumFlow = Constraint(
        model.s_O,
        model.s_T,
        rule=BeneficialReuseMinimumFlowRule,
        doc="Beneficial reuse minimum flow when selected",
    )